shell/platform/embedder/tests/embedder_unittests.cc

```diff
@@ -1308,6 +1536,156 @@ TEST_F(EmbedderTest, CannotRemoveUnknownView) {
+  add_duplicate_view_info.user_data = &captures;
+  add_duplicate_view_info.add_view_callback =
+      [](const FlutterAddViewResult* result) {
+        auto captures = reinterpret_cast<Captures*>(result->user_data);
+
+        ASSERT_FALSE(result->added);
+        ASSERT_TRUE(captures->add_first_view.IsSignaledForTest());
+        ASSERT_FALSE(captures->add_duplicate_view.IsSignaledForTest());
+        ASSERT_FALSE(captures->add_second_view.IsSignaledForTest());
+        ASSERT_FALSE(captures->remove_second_view.IsSignaledForTest());
+
+        captures->add_duplicate_view.Signal();
+      };
+
+  // Add view 456.
+  FlutterAddViewInfo add_second_view_info = {};
+  add_second_view_info.struct_size = sizeof(FlutterAddViewInfo);
+  add_second_view_info.view_id = 456;
+  add_second_view_info.view_metrics = &metrics456;
+  add_second_view_info.user_data = &captures;
+  add_second_view_info.add_view_callback =
+      [](const FlutterAddViewResult* result) {
+        auto captures = reinterpret_cast<Captures*>(result->user_data);
+
+        ASSERT_TRUE(result->added);
+        ASSERT_TRUE(captures->add_first_view.IsSignaledForTest());
+        ASSERT_TRUE(captures->add_duplicate_view.IsSignaledForTest());
+        ASSERT_FALSE(captures->add_second_view.IsSignaledForTest());
+        ASSERT_FALSE(captures->remove_second_view.IsSignaledForTest());
+
+        captures->add_second_view.Signal();
+      };
+
+  // Remove view 456.
+  FlutterRemoveViewInfo remove_second_view_info = {};
+  remove_second_view_info.struct_size = sizeof(FlutterRemoveViewInfo);
+  remove_second_view_info.view_id = 456;
+  remove_second_view_info.user_data = &captures;
+  remove_second_view_info.remove_view_callback =
+      [](const FlutterRemoveViewResult* result) {
+        auto captures = reinterpret_cast<Captures*>(result->user_data);
+
+        ASSERT_TRUE(result->removed);
+        ASSERT_TRUE(captures->add_first_view.IsSignaledForTest());
+        ASSERT_TRUE(captures->add_duplicate_view.IsSignaledForTest());
+        ASSERT_TRUE(captures->add_second_view.IsSignaledForTest());
+        ASSERT_FALSE(captures->remove_second_view.IsSignaledForTest());
+
+        captures->remove_second_view.Signal();
+      };
+
+  ASSERT_EQ(FlutterEngineAddView(engine.get(), &add_view_info), kSuccess);
+  ASSERT_EQ(FlutterEngineAddView(engine.get(), &add_duplicate_view_info),
+            kSuccess);
+  ASSERT_EQ(FlutterEngineAddView(engine.get(), &add_second_view_info),
+            kSuccess);
+  ASSERT_EQ(FlutterEngineRemoveView(engine.get(), &remove_second_view_info),
+            kSuccess);
+  captures.remove_second_view.Wait();
+  captures.add_second_view.Wait();
+  captures.add_duplicate_view.Wait();
+  captures.add_first_view.Wait();
+  ASSERT_EQ(message_count, 3);
+}
+
+TEST_F(EmbedderTest, CanRenderMultipleViews) {
+  // TODO(dkwingsmt): Support rendering views that aren't the implicit view.
+  // https://github.com/flutter/flutter/issues/135530
```

Comment:
@dkwingsmt The engine crashes if you render a view that isn't the implicit view: [`embedder_external_view_embedder.cc#L65-L68`](https://github.com/flutter/engine/blob/7ea6b003ffeb5113b709e8b336f5e6ae22e39a38/shell/platform/embedder/embedder_external_view_embedder.cc#L65-L68)

```
[FATAL:flutter/shell/platform/embedder/embedder_external_view_embedder.cc(68)] Check failed: flutter_view_id == kFlutterImplicitViewId.
```

Here was my C++ test:

```cpp
TEST_F(EmbedderTest, CanRenderMultipleViews) {
  auto& context = GetEmbedderContext(EmbedderTestContextType::kSoftwareContext);
  EmbedderConfigBuilder builder(context);
  builder.SetSoftwareRendererConfig();
  builder.SetCompositor();
  builder.SetDartEntrypoint("render_all_views");

  fml::AutoResetWaitableEvent latch0, latch123;
  context.GetCompositor().SetPresentCallback(
      [&](FlutterViewId view_id, const FlutterLayer** layers,
          size_t layers_count) {
        switch (view_id) {
          case 0:
            latch0.Signal();
            break;
          case 123:
            latch123.Signal();
            break;
          default:
            FML_UNREACHABLE();
        }
      },
      /* one_shot= */ false);

  auto engine = builder.LaunchEngine();
  ASSERT_TRUE(engine.is_valid());

  FlutterWindowMetricsEvent metrics = {};
  metrics.struct_size = sizeof(FlutterWindowMetricsEvent);
  metrics.width = 800;
  metrics.height = 600;
  metrics.pixel_ratio = 1.0;
  metrics.view_id = 123;

  FlutterAddViewInfo add_view_info = {};
  add_view_info.struct_size = sizeof(FlutterAddViewInfo);
  add_view_info.view_id = 123;
  add_view_info.view_metrics = &metrics;
  add_view_info.add_view_callback = [](const FlutterAddViewResult* result) {
    ASSERT_TRUE(result->added);
  };

  ASSERT_EQ(FlutterEngineAddView(engine.get(), &add_view_info), kSuccess);

  latch0.Wait();
  latch123.Wait();
}
```

Here was my Dart entrypoint:

```dart
@pragma('vm:entry-point')
void render_all_views() {
  PlatformDispatcher.instance.onBeginFrame = (Duration duration) {
    for (final FlutterView view in PlatformDispatcher.instance.views) {
      final Size size = Size(800.0, 600.0);
      final Color red = Color.fromARGB(127, 255, 0, 0);

      final SceneBuilder builder = SceneBuilder();

      builder.pushOffset(0.0, 0.0);

      builder.addPicture(
          Offset(0.0, 0.0), CreateColoredBox(red, size));

      builder.pop();

      view.render(builder.build());
    }
  };
}
```

Reply:
Found a solution, closing!